File: python/binary-search/binary_search.py
"""Binary search."""


def find(search_list: list[int], value: int) -> int:
    """Return the index of a value in a list, using binary search."""
    lower = 0
    upper = len(search_list)
    while lower != upper:
        # Take the center point as the pivot.
        pivot = (upper + lower) // 2
        if search_list[pivot] == value:
            return pivot
        # Reduce the search to the upper/lower half of the list.
        if search_list[pivot] > value:
            upper = pivot
        else:
            lower = pivot + 1
    raise ValueError("value not in array")

File: go/binary-search/binary_search.go
package binarysearch

// SearchInts returns the index of an int in a list using a binary search.
func SearchInts(list []int, key int) int {
	for low, high := 0, len(list); low != high; {
		mid := (high + low) / 2
		val := list[mid]
		switch {
		case val == key:
			return mid
		case val > key:
			high = mid
		default:
			low = mid + 1
		}
	}

	return -1
}
